src/vs/workbench/common/editor/editorLabels.ts

```diff
@@ -0,0 +1,188 @@
+				const oldEnablement = this.enabled;
+				this.storeEnablementState();
+				if (oldEnablement !== this.enabled && this.patterns.length > 0) {
+					this._onDidChange.fire();
+				}
+			}
+
+			// Cache the patterns
+			else if (e.affectsConfiguration(CustomEditorLabelService.SETTING_ID_PATTERNS)) {
+				this.storeCustomPatterns();
+				this._onDidChange.fire();
+			}
+		}));
+	}
+
+	private storeEnablementState(): void {
+		this.enabled = this.configurationService.getValue<boolean>(CustomEditorLabelService.SETTING_ID_ENABLED);
+	}
+
+	private storeCustomPatterns(): void {
+		this.patterns = [];
+		const customLabelPatterns = this.configurationService.getValue<IEditorCustomLabelObject>(CustomEditorLabelService.SETTING_ID_PATTERNS);
+		for (const pattern in customLabelPatterns) {
+			const template = customLabelPatterns[pattern];
+			this.patterns.push({ pattern, template });
+		}
+
+		this.patterns.sort((a, b) => this.patternWeight(b.pattern) - this.patternWeight(a.pattern));
+	}
+
+	private patternWeight(pattern: string): number {
+		pattern = pattern.replace('\\\\', '/');
+
+		let weight = 0;
+		for (const fragment of pattern.split('/')) {
+			if (fragment === '**') {
+				weight += 1;
+			} else if (fragment === '*') {
+				weight += 10;
+			} else if (fragment.includes('*') || fragment.includes('?')) {
+				weight += 50;
+			} else if (fragment !== '') {
+				weight += 100;
+			}
+		}
+
+		return weight;
+	}
+
+	getName(resource: URI): string | undefined {
+		if (!this.enabled) {
+			return undefined;
+		}
+		return this.applyPatterns(resource);
+	}
+
+	private applyPatterns(resource: URI): string | undefined {
+		const root = this.workspaceContextService.getWorkspaceFolder(resource);
+		for (const { pattern, template } of this.patterns) {
+			if (this.matchPattern(root, resource, pattern)) {
+				const attrs = this.getFileAttributes(resource);
+				return this.applyTempate(template, attrs);
+			}
+		}
+		return undefined;
+	}
+
+	private matchPattern(root: IWorkspaceFolder | null, resource: URI, pattern: string): boolean {
+		const patternURI = URI.from({ scheme: 'glob', path: pattern }); // Used to figure out if the pattern is absolute
+		const relevantPath = !root || isAbsolutePath(patternURI) ? resource.fsPath : relativePath(root.uri, resource) ?? resource.fsPath;
+
+		return match(pattern, relevantPath);
+	}
+
+	private getFileAttributes(resource: URI): IFilenameAttributes {
+		// grand-parent/parent/filename.ext1.ext2 -> parent
+		const directoryURI = dirname(resource);
+		const directoryName = basename(directoryURI);
+
+		// grand-parent/parent/filename.ext1.ext2 -> [grand-parent, parent]
+		const pathFragments = directoryURI.path.split('/');
+
+		// grand-parent/parent/filename.ext1.ext2 -> filename.ext1
+		const base = basename(resource);
+		const baseSplit = base.split('.');
+		const filename = baseSplit.length > 1 ? baseSplit.slice(0, -1).join('.') : baseSplit[0];
```

Comment:
Why would this not be `basename.extname`?

Reply:
Gives more control